packages/studio-be/src/core/migration/bot-migration-service.ts

```diff
@@ -68,58 +58,66 @@ export class BotMigrationService {
     await this.bpfs.forBot(botId).upsertFile('/', 'migrations.json', JSON.stringify(entries, undefined, 2))
   }
 
-  private displayMigrationStatus(configVersion: string, missingMigrations: MigrationFile[], logger: sdk.Logger) {
-    const migrations = missingMigrations.map(x => this.migService.loadedMigrations[x.filename].info)
+  private async getMissingMigrations(botId: string, migrations: MigrationFile[]): Promise<MigrationFile[]> {
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    const opts = await this.migService.getMigrationOpts({ botId, botConfig, isDryRun: true })
+
+    const missing = await Promise.mapSeries(migrations, async migration => {
+      try {
+        const result = await this.migService.loadedMigrations[migration.filename].up(opts)
+        if (result.hasChanges) {
+          return migration
+        }
+      } catch (err) {
+        return migration
+      }
+    })
 
-    logger.warn(chalk`
-${_.repeat(' ', 9)}========================================
-{bold ${centerText(`Migration${migrations.length === 1 ? '' : 's'} Required`, 40, 9)}}
-{dim ${centerText(`Version ${configVersion} => ${this.migService.targetVersion} `, 40, 9)}}
-{dim ${centerText(`${migrations.length} change${migrations.length === 1 ? '' : 's'}`, 40, 9)}}
-${_.repeat(' ', 9)}========================================`)
+    return missing.filter(x => x !== undefined) as MigrationFile[]
```

Comment:
PP: Instead of filtering undefined values from the list of missing migrations, you could simply use a for loop and an array of MigrationFile.

Reply:
I'm not sure to get your suggestion. Basically that loop tests all migrations and if it would make changes, then it adds it to the array. Most of the time that array will be filled with undefined values, I prefer to clear out the values before returning the array

Reply:
I mean instead of iterating twice on the migration list, you could do it only once. I personally dislike looping over an array to then remove unwanted data afterward. Makes more sense to do it in one shot.